dvc/utils/strictyaml.py

```diff
@@ -141,49 +141,58 @@ def prepare_code(mark: "StreamMark") -> "Syntax":
-    ['stages', 'metrics'], metrics being a boolean type does not have `lc`
-    prop.
-    ```
-    stages:
-      metrics: true
-    ```
-
-    To provide some context to the user, we step up to the
-    path ['stages'], which being a collection type, will have `lc` prop
-    with which we can find line and col.
-
-    This may end up being not accurate, so we try to show the same amount of
-    lines of code for `n` number of steps taken upwards. In a worst case,
-    it may be just 1 step (as non-collection item cannot have child items),
-    but `schema validator` may provide us arbitrary path. So, this caps the
-    number of steps upward to just 5. If it does not find any linecols, it'll
-    abort.
+def _normalize_linecol(lc: "LineCol | Tuple[Any, Any]") -> Tuple[int, int]:
+    from ruamel.yaml.comments import LineCol
+
+    line, col = None, None
+
+    if isinstance(lc, LineCol):
+        line = lc.line
+        col = lc.col
+    elif isinstance(lc, tuple):
+        line = int(lc[0])
+        col = int(lc[1])
+
+    if not (isinstance(line, int) and isinstance(col, int)):
+        raise TypeError(f"Unable to determine neither line nor col ({line}:{col})")
+
+    return line + 1, col + 1
+
+
+def determine_linecol(data: Any, location: "Glob") -> Tuple[int, int]:
+    """
+    Return the line and column number for the given location in the data.
+
+    Args:
+        data: The data must be a parsed YAML document represented as a Python object,
+            obtained by calling `ruamel.yaml.YAML(typ='rt').load()` on a YAML file.
+            The function expects the data to contain information about the original YAML
+            document's structure, including the line and column numbers where each
+            element appears.
+        location: The glob pattern to match a key or item in the data.
+
+    Raises:
+        ValueError: Raised if more than one leaf matches the glob.
+        KeyError: Raised if the location not found.
+
+    Returns:
+        A tuple containing the line and column number of the matched location.
     """
     from dpath import get
+    from ruamel.yaml.comments import CommentedMap, CommentedSeq
 
-    step = 1
-    line, col = None, None
-    while paths and step < max_steps:
-        value = get(data, paths, default=None)
-        if value is not None:
-            with suppress(AttributeError, TypeError):
-                line = value.lc.line + 1  # type: ignore[attr-defined]
-                col = value.lc.col + 1  # type: ignore[attr-defined]
-                break
-        step += 1
-        *paths, _ = paths
+    if isinstance((obj := get(data, location)), (CommentedSeq, CommentedMap)):
+        return _normalize_linecol(obj.lc)
+
+    obj = get(data, location[:-1])
+    if isinstance(obj, CommentedMap):
+        return _normalize_linecol(obj.lc.key(location[-1]))
+    if isinstance(obj, CommentedSeq):
+        return _normalize_linecol(obj.lc.item(int(location[-1])))
```

Comment:
I didn't know about `lc.key`/`lc.value`. Looks like `lc.key` is for `key` part only, and `lc.value` is needed for value part.

So, only the first line will be shown for following yaml file here, whereas users may be interested on seeing both of the lines. (`lc.value` is still the starting position, it's too bad that we don't have end position for values, otherwise it's almost perfect.)

```yaml
outs:
 models/
```

Reply:
> I didn't know about `lc.key`/`lc.value`. Looks like `lc.key` is for `key` part only, and `lc.value` is needed for value part.
> 
> So, only the first line will be shown for following yaml file here, whereas users may be interested on seeing both of the lines. (`lc.value` is still the starting position of the value, it's too bad that we don't have end position for values, otherwise it's almost perfect.)
> 
> ```yaml
> outs:
>  models/
> ```

Thank you for bringing this to my attention. I'll commit a fix for this issue soon, which will allow both lines to be displayed correctly.